### PR TITLE
fix(NODE-3150): allow retrieving PCRE-style RegExp

### PR DESCRIPTION
**Pull request by andymina**
## Description
- Allowed retrieving PCRE-style RegExp by adding `bsonRegExp` option to commands, collections, and dbs
- Added unit tests for the `bsonRegExp` option

### CODE REVIEW

**Comment by nbbeeken** [comment]
I don't have further background on this, we have a precedent of other values being auto converted with the `promoteValues` and like settings. However, I know it is a strict requirement that BSON libraries have an API that explicitly converts UUIDs to native types, so that is a precedence for the opposite (as well as the regex examples from other libs you mention). There is convenience in handing the user the native type, up until it fails because of the contents.

Maybe as a separate effort we should take a look at the BSON library catching a Regexp decoding error and using BSONRegex automatically instead of failing, although I imagine that would also need a flag.

**Comment by durran** [comment]
Yeah it's a weird one since it is backwards breaking anyways so I guess we are stuck with a flag. Thanks for the clarification.

File: src/connection_string.ts
Comment (durran) [inline src/connection_string.ts:591]: I'm not sure we want this to be a connection string option as it's Node driver specific.

**Review by durran — CHANGES_REQUESTED** [review]

**Comment by durran** [comment]
I would also change the description not to mention `findOne` specifically, as the option is set on the command response before deserialization, which will actually take care of all commands (not just `findOne`).

File: src/connection_string.ts
Comment (andymina) [reply src/connection_string.ts:591]: I included the new bsonRegExp option here because this is where all possible options (uri, object param, and defaults) are defined so it might be a little overreaching to solve this issue now. I can create a new ticket to make sure you can’t use it in the connection string if that sounds better?

Comment (nbbeeken) [reply src/connection_string.ts:591]: Capturing some investigation from a slack convo:
Andy discovered we do support BSON options in the URI in 3.6 so the behavior is consistent, may be unusual, not sure what other drivers do but I think we can certainly defer changing that to another ticket.

File: test/unit/bson_regex.test.js
Comment (nbbeeken) [inline test/unit/bson_regex.test.js:16]: We should use a better db/collection name here

File: test/unit/bson_regex.test.js
Comment (nbbeeken) [inline test/unit/bson_regex.test.js:35]: The setup is nearly the same for each test, can we share some of this logic in a before / beforeEach.
Or I think we can make a loop that tests each of these, with only one implementation:

```javascript
const options = { bsonRegexp: true };
for (const passOptionsTo of ['client', 'db',  'collection', 'operation']) {
  it(`should respond with BSONRegExp class with option passed to ${passOptionsTo}`, async function () {
        client = this.configuration.newClient(passOptionsTo === 'client' ? options : undefined);
        await client.connect();

        const db = client.db('a', passOptionsTo === 'db' ? options : undefined);
        const collection = db.collection('b', passOptionsTo === 'collection' ? options : undefined);

        await collection.insertOne({ regex: new BSONRegExp('abc', 'imx') });
        const res = await collection.findOne(
          { regex: new BSONRegExp('abc', 'imx') },
          passOptionsTo === 'operation' ? options : undefined
        );
        
        // ....
  });
}
```
How does this look? needs some clean up but this is the idea.

**Review by nbbeeken — CHANGES_REQUESTED** [review]
Some test clean up requests, driver changes LGTM

File: test/unit/bson_regex.test.js
Comment (durran) [reply test/unit/bson_regex.test.js:35]: Looks good to me.

File: src/connection_string.ts
Comment (durran) [reply src/connection_string.ts:591]: I'm fine with that.

**Review by durran — APPROVED** [review]
I'm LGTM once @nbbeeken 's test change requests are in.

File: src/sdam/server.ts
Comment (nbbeeken) [inline src/sdam/server.ts:85]: Looks like this got brought back in the rebase, I think this is removed now.

**Review by nbbeeken — APPROVED** [review]

**Review by dariakp — APPROVED** [review]
LGTM

**Review by emadum — APPROVED** [review]
LGTM

**Pull request merged**
